cmd/cmd.go

```diff
@@ -33,3 +51,221 @@ const msgMissingMetadata = `
+	}
+
+	if res.StatusCode == http.StatusUnauthorized {
+		siteURL := config.InferSiteURL(ctx.usrCfg.GetString("apibaseurl"))
+		return fmt.Errorf("unauthorized request. Please run the configure command. You can find your API token at %s/my/settings", siteURL)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		switch ctx.payload.Error.Type {
+		case "track_ambiguous":
+			return fmt.Errorf("%s: %s", ctx.payload.Error.Message, strings.Join(ctx.payload.Error.PossibleTrackIDs, ", "))
+		default:
+			return errors.New(ctx.payload.Error.Message)
+		}
+	}
+	return nil
+}
+
+func (d *downloadContext) writeMetadata() error {
+	if err := d.validate(); err != nil {
+		return err
+	}
+	metadata := d.getMetadata()
+	exercise := d.getExercise()
+	if err := metadata.Write(exercise.MetadataDir()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *downloadContext) writeSolutionFiles() error {
+	if err := d.validate(); err != nil {
+		return err
+	}
+	exercise := d.getExercise()
+	for _, file := range d.payload.Solution.Files {
+		unparsedURL := fmt.Sprintf("%s%s", d.payload.Solution.FileDownloadBaseURL, file)
+		parsedURL, err := netURL.ParseRequestURI(unparsedURL)
+		if err != nil {
+			return err
+		}
+
+		client, err := api.NewClient(d.usrCfg.GetString("token"), d.usrCfg.GetString("apibaseurl"))
+		req, err := client.NewRequest("GET", parsedURL.String(), nil)
+		if err != nil {
+			return err
+		}
+
+		res, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer res.Body.Close()
+
+		if res.StatusCode != http.StatusOK {
+			// TODO: deal with it
+			continue
+		}
+		// Don't bother with empty files.
+		if res.Header.Get("Content-Length") == "0" {
+			continue
+		}
+
+		// TODO: if there's a collision, interactively resolve (show diff, ask if overwrite).
+		// TODO: handle --force flag to overwrite without asking.
+
+		// Work around a path bug due to an early design decision (later reversed) to
+		// allow numeric suffixes for exercise directories, allowing people to have
+		// multiple parallel versions of an exercise.
+		pattern := fmt.Sprintf(`\A.*[/\\]%s-\d*/`, exercise.Slug)
+		rgxNumericSuffix := regexp.MustCompile(pattern)
+		if rgxNumericSuffix.MatchString(file) {
+			file = string(rgxNumericSuffix.ReplaceAll([]byte(file), []byte("")))
+		}
+
+		// Rewrite paths submitted with an older, buggy client where the Windows path is being treated as part of the filename.
+		file = strings.Replace(file, "\\", "/", -1)
+
+		relativePath := filepath.FromSlash(file)
+
+		dir := filepath.Join(exercise.MetadataDir(), filepath.Dir(relativePath))
+		if err := os.MkdirAll(dir, os.FileMode(0755)); err != nil {
+			return err
+		}
+
+		f, err := os.Create(filepath.Join(exercise.MetadataDir(), relativePath))
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if _, err := io.Copy(f, res.Body); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *downloadContext) getExercise() workspace.Exercise {
```

Comment:
Style-wise in the CLI we typically don't use `get` prefixes for getters, we'll usually name the method after what it returns (so `exercise()`).